File: android/app/src/main/java/com/example/dinochase/GameState.kt
package com.example.dinochase

import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * Observable game state shared between the Compose UI (menus / HUD) and the
 * [GameEngine]. The engine mutates score / time / combo as play unfolds; the UI
 * reads them to render overlays. Mirrors the iOS `GameState`.
 */
class GameState(
    private val haptics: Haptics,
    private val prefs: SharedPreferences,
) {
    enum class Phase { MENU, PLAYING, PAUSED, GAME_OVER }

    /** Difficulty tunes round length and monkey speed / aggression. */
    enum class Difficulty(
        val label: String,
        val emoji: String,
        val startingTime: Double,
        val monkeyBaseSpeed: Float,
        val speedRamp: Float,
    ) {
        EASY("Easy", "🌱", 30.0, 250f, 18f),
        NORMAL("Normal", "🔥", 20.0, 300f, 26f),
        HARD("Hard", "💀", 15.0, 360f, 34f),
    }

    var phase by mutableStateOf(Phase.MENU)
    var score by mutableStateOf(0)
    var timeRemaining by mutableStateOf(20.0)
    var highScore by mutableStateOf(prefs.getInt(KEY_HIGH, 0))
        private set

    var comboMultiplier by mutableStateOf(1)
    var comboTimeRemaining by mutableStateOf(0.0)
    val comboWindow = 2.5

    var roarCharge by mutableStateOf(0.0)   // 0..1
    val roarReady: Boolean get() = roarCharge >= 1.0

    var flashMessage by mutableStateOf<String?>(null)

    var difficulty by mutableStateOf(
        Difficulty.entries.getOrElse(prefs.getInt(KEY_DIFF, Difficulty.NORMAL.ordinal)) { Difficulty.NORMAL }
    )
        private set

    /** Wired up by [GameEngine] once both objects exist. */
    lateinit var engine: GameEngine

    private val bonusPerCatch = 3.0
    private val bananaBonus = 2.0
    private val catchesPerRoar = 3.0
    val roarDuration = 3.0

    private var flashTimer = 0.0

    // MARK: - Menu / difficulty

    fun chooseDifficulty(d: Difficulty) {
        difficulty = d
        prefs.edit().putInt(KEY_DIFF, d.ordinal).apply()
    }

    // MARK: - Phase transitions

    fun start() {
        score = 0
        timeRemaining = difficulty.startingTime
        comboMultiplier = 1
        comboTimeRemaining = 0.0
        roarCharge = 0.0
        flashMessage = null
        flashTimer = 0.0
        phase = Phase.PLAYING
        engine.startNewGame()
    }

    fun toMenu() {
        phase = Phase.MENU
        engine.resetToIdle()
    }

    fun pause() {
        if (phase != Phase.PLAYING) return
        phase = Phase.PAUSED
        engine.paused = true
    }

    fun resume() {
        if (phase != Phase.PAUSED) return
        phase = Phase.PLAYING
        engine.paused = false
    }

    fun roar() {
        if (phase != Phase.PLAYING || !roarReady) return
        roarCharge = 0.0
        haptics.roar()
        flash("🦖 ROAAAR!")
        engine.triggerRoar(roarDuration)
    }

    // MARK: - Scoring (called from the engine)

    fun caughtMonkey() {
        comboMultiplier = if (comboTimeRemaining > 0) (comboMultiplier + 1).coerceAtMost(9) else 1
        comboTimeRemaining = comboWindow

        score += comboMultiplier
        timeRemaining += bonusPerCatch
        roarCharge = (roarCharge + 1.0 / catchesPerRoar).coerceAtMost(1.0)

        haptics.catchThump(comboMultiplier)
        flash(if (comboMultiplier >= 2) "x$comboMultiplier!  +$comboMultiplier" else "+${bonusPerCatch.clean()}s")
    }

    fun collectedBanana() {
        score += 1
        timeRemaining += bananaBonus
        haptics.tick()
        flash("🍌 +${bananaBonus.clean()}s")
    }

    // MARK: - Ticking

    fun tick(dt: Double) {
        if (phase != Phase.PLAYING) return

        timeRemaining -= dt
        if (comboTimeRemaining > 0) {
            comboTimeRemaining -= dt
            if (comboTimeRemaining <= 0) {
                comboTimeRemaining = 0.0
                comboMultiplier = 1
            }
        }
        if (flashTimer > 0) {
            flashTimer -= dt
            if (flashTimer <= 0) flashMessage = null
        }

        if (timeRemaining <= 0) {
            timeRemaining = 0.0
            endGame()
        }
    }

    private fun flash(text: String) {
        flashMessage = text
        flashTimer = 0.9
    }

    private fun endGame() {
        if (score > highScore) {
            highScore = score
            prefs.edit().putInt(KEY_HIGH, highScore).apply()
        }
        haptics.gameOver()
        phase = Phase.GAME_OVER
        engine.stopGame()
    }

    private fun Double.clean(): String =
        if (this == kotlin.math.floor(this)) toInt().toString() else "%.1f".format(this)

    companion object {
        private const val KEY_HIGH = "DinoChaseHighScore"
        private const val KEY_DIFF = "DinoChaseDifficulty"
    }
}

File: DinoChase/DinoChase/GameState.swift
import SwiftUI
import SpriteKit

/// Shared, observable game state that bridges SwiftUI (menus / HUD) and the
/// SpriteKit `GameScene`. The scene mutates `score` / `timeRemaining`; SwiftUI
/// reads them to render overlays.
final class GameState: ObservableObject {
    enum Phase {
        case menu
        case playing
        case paused
        case gameOver
    }

    @Published var phase: Phase = .menu
    @Published var score: Int = 0
    @Published var timeRemaining: Double = 20
    @Published private(set) var highScore: Int = UserDefaults.standard.integer(forKey: "DinoChaseHighScore")

    // Combo / streak: catch monkeys in quick succession to build a multiplier.
    @Published var comboMultiplier: Int = 1
    @Published var comboTimeRemaining: Double = 0
    /// Time window (seconds) to land the next catch before the combo resets.
    let comboWindow: Double = 2.5

    /// Transient banner text (e.g. "+3s", "x4!") shown briefly in the HUD.
    @Published var flashMessage: String? = nil

    /// The single persistent scene instance, retained here so it survives
    /// SwiftUI view redraws.
    var scene: GameScene?

    /// Seconds granted per round at the start.
    let startingTime: Double = 20
    /// Base seconds added to the clock for each monkey caught.
    let bonusPerCatch: Double = 3
    /// Seconds added when a banana is collected.
    let bananaBonus: Double = 2

    private var flashClearWorkItem: DispatchWorkItem?

    // MARK: - Phase transitions

    func start() {
        score = 0
        timeRemaining = startingTime
        comboMultiplier = 1
        comboTimeRemaining = 0
        flashMessage = nil
        phase = .playing
        scene?.startNewGame()
    }

    func toMenu() {
        phase = .menu
        scene?.resetToIdle()
    }

    func pause() {
        guard phase == .playing else { return }
        phase = .paused
        scene?.setPaused(true)
    }

    func resume() {
        guard phase == .paused else { return }
        phase = .playing
        scene?.setPaused(false)
    }

    // MARK: - Scoring

    func caughtMonkey() {
        // Extend the combo if still within the window; otherwise start fresh.
        if comboTimeRemaining > 0 {
            comboMultiplier = min(comboMultiplier + 1, 9)
        } else {
            comboMultiplier = 1
        }
        comboTimeRemaining = comboWindow

        score += comboMultiplier
        timeRemaining += bonusPerCatch

        Haptics.shared.catchThump(intensity: comboMultiplier)

        if comboMultiplier >= 2 {
            flash("x\(comboMultiplier)!  +\(comboMultiplier)")
        } else {
            flash("+\(bonusPerCatch.clean)s")
        }
    }

    func collectedBanana() {
        score += 1
        timeRemaining += bananaBonus
        Haptics.shared.tick()
        flash("🍌 +\(bananaBonus.clean)s")
    }

    // MARK: - Ticking

    func tick(_ delta: Double) {
        guard phase == .playing else { return }

        timeRemaining -= delta
        if comboTimeRemaining > 0 {
            comboTimeRemaining -= delta
            if comboTimeRemaining <= 0 {
                comboTimeRemaining = 0
                comboMultiplier = 1
            }
        }

        if timeRemaining <= 0 {
            timeRemaining = 0
            endGame()
        }
    }

    // MARK: - Helpers

    private func flash(_ text: String) {
        flashMessage = text
        flashClearWorkItem?.cancel()
        let work = DispatchWorkItem { [weak self] in self?.flashMessage = nil }
        flashClearWorkItem = work
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.9, execute: work)
    }

    private func endGame() {
        if score > highScore {
            highScore = score
            UserDefaults.standard.set(highScore, forKey: "DinoChaseHighScore")
        }
        Haptics.shared.gameOver()
        phase = .gameOver
        scene?.stopGame()
    }
}

private extension Double {
    /// "3" not "3.0" — for tidy HUD strings.
    var clean: String {
        self == rounded() ? String(Int(self)) : String(format: "%.1f", self)
    }
}
